File: src/test/java/net/explorviz/adapter/service/converter/AttributesReaderTest.java
package net.explorviz.adapter.service.converter;

import static org.junit.jupiter.api.Assertions.assertEquals;

import com.google.protobuf.ByteString;
import io.opentelemetry.proto.common.v1.AnyValue;
import io.opentelemetry.proto.common.v1.KeyValue;
import io.opentelemetry.proto.trace.v1.Span;
import java.nio.charset.Charset;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import net.explorviz.adapter.service.validation.StrictValidator;
import org.junit.jupiter.api.Test;

public class AttributesReaderTest {

  private static final String KEY_LANDSCAPE_TOKEN = AttributesReader.LANDSCAPE_TOKEN;
  private static final String KEY_LANDSCAPE_SECRET = AttributesReader.TOKEN_SECRET;
  private static final String KEY_HOST_NAME = AttributesReader.HOST_NAME;
  private static final String KEY_HOST_IP = AttributesReader.HOST_IP;
  private static final String KEY_APPLICATION_NAME = AttributesReader.APPLICATION_NAME;
  private static final String KEY_APPLICATION_INSTANCE_ID =
      AttributesReader.APPLICATION_INSTANCE_ID;
  private static final String KEY_APPLICATION_LANGUAGE = AttributesReader.APPLICATION_LANGUAGE;
  private static final String KEY_METHOD_FQN = AttributesReader.METHOD_FQN;

  private static final String TOKEN = "tok";
  private static final String SECRET = "secret";
  private static final String HOSTNAME = "Host";
  private static final String HOST_IP = "1.2.3.4";
  private static final String APP_NAME = "Test App";
  private static final String APP_INSTANCE_ID = "1234L";
  private static final String APP_LANG = "java";
  private static final String FQN = "foo.bar.test()";
  private static final String K8S_POD_NAME = "my name is pod. james pod.";
  private static final String K8S_NAMESPACE = "name ";
  private static final String K8S_DEPLOYMENT = "iraq. i committed war crimes there, ordered by the US government.";
  private static final String K8S_NODE = "node.js";




  @Test
  void testValidSpanReadOut() {
    List<KeyValue> attr = generateValidAttributesMap();
    Span span = generateSpanFromAttributesMap(attr);
    final AttributesReader reader = new AttributesReader(span);

    assertExcept(reader, null);
    
  }

  @Test
  void testDefaultTokenReadOut() {
    List<KeyValue> attr = generateValidAttributesMap();
    attr = removeElementAndReturnAttributesMap(KEY_LANDSCAPE_TOKEN, attr);
    Span span = generateSpanFromAttributesMap(attr);
    final AttributesReader reader = new AttributesReader(span);
    
    assertExcept(reader, AttributesReader.LANDSCAPE_TOKEN);
    assertEquals(reader.getLandscapeToken(), DefaultAttributeValues.DEFAULT_LANDSCAPE_TOKEN);
  }

  @Test
  void testDefaultSecretReadOut() {
    List<KeyValue> attr = generateValidAttributesMap();
    attr = removeElementAndReturnAttributesMap(KEY_LANDSCAPE_SECRET, attr);
    Span span = generateSpanFromAttributesMap(attr);
    final AttributesReader reader = new AttributesReader(span);
    
    assertExcept(reader, AttributesReader.TOKEN_SECRET);
    assertEquals(reader.getSecret(), DefaultAttributeValues.DEFAULT_LANDSCAPE_SECRET);
  }

  @Test
  void testDefaultHostnameReadOut() {
    List<KeyValue> attr = generateValidAttributesMap();
    attr = removeElementAndReturnAttributesMap(KEY_HOST_NAME, attr);
    Span span = generateSpanFromAttributesMap(attr);
    final AttributesReader reader = new AttributesReader(span);

    assertExcept(reader, AttributesReader.HOST_NAME);
    assertEquals(reader.getHostName(), DefaultAttributeValues.DEFAULT_HOST_NAME);
  }

  @Test
  void testDefaultHostIpReadOut() {
    List<KeyValue> attr = generateValidAttributesMap();
    attr = removeElementAndReturnAttributesMap(KEY_HOST_IP, attr);
    Span span = generateSpanFromAttributesMap(attr);
    final AttributesReader reader = new AttributesReader(span);

    assertExcept(reader, AttributesReader.HOST_IP);
    assertEquals(reader.getHostIpAddress(), DefaultAttributeValues.DEFAULT_HOST_IP);
  }

  @Test
  void testDefaultAppNameReadOut() {
    List<KeyValue> attr = generateValidAttributesMap();
    attr = removeElementAndReturnAttributesMap(KEY_APPLICATION_NAME, attr);
    Span span = generateSpanFromAttributesMap(attr);
    final AttributesReader reader = new AttributesReader(span);

    assertExcept(reader, AttributesReader.APPLICATION_NAME);
    assertEquals(reader.getApplicationName(), DefaultAttributeValues.DEFAULT_APP_NAME);
  }

  @Test
  void testDefaultAppInstanceIdReadOut() {
    List<KeyValue> attr = generateValidAttributesMap();
    attr = removeElementAndReturnAttributesMap(KEY_APPLICATION_INSTANCE_ID, attr);
    Span span = generateSpanFromAttributesMap(attr);
    final AttributesReader reader = new AttributesReader(span);

    assertExcept(reader, AttributesReader.APPLICATION_INSTANCE_ID);
    assertEquals(reader.getApplicationInstanceId(), DefaultAttributeValues.DEFAULT_APP_INSTANCE_ID);
  }

  @Test
  void testDefaultAppLangReadOut() {
    List<KeyValue> attr = generateValidAttributesMap();
    attr = removeElementAndReturnAttributesMap(KEY_APPLICATION_LANGUAGE, attr);
    Span span = generateSpanFromAttributesMap(attr);
    final AttributesReader reader = new AttributesReader(span);

    assertExcept(reader, AttributesReader.APPLICATION_LANGUAGE);
    assertEquals(reader.getApplicationLanguage(), DefaultAttributeValues.DEFAULT_APP_LANG);
  }


  @Test
  void testDefaultFqnReadOut() {
    List<KeyValue> attr = generateValidAttributesMap();
    attr = removeElementAndReturnAttributesMap(KEY_METHOD_FQN, attr);
    Span span = generateSpanFromAttributesMap(attr);
    final AttributesReader reader = new AttributesReader(span);

    assertExcept(reader, AttributesReader.METHOD_FQN);
    assertEquals(reader.getMethodFqn(), DefaultAttributeValues.DEFAULT_FQN);
  }
  
  void assertExcept(AttributesReader reader, String except){
    if(!Objects.equals(except, AttributesReader.LANDSCAPE_TOKEN))
        assertEquals(reader.getLandscapeToken(), TOKEN);
    if(!Objects.equals(except, AttributesReader.TOKEN_SECRET))
        assertEquals(reader.getSecret(), SECRET);
    if(!Objects.equals(except, AttributesReader.HOST_NAME))
        assertEquals(reader.getHostName(), HOSTNAME);
    if(!Objects.equals(except, AttributesReader.HOST_IP))
        assertEquals(reader.getHostIpAddress(), HOST_IP);
    if(!Objects.equals(except, AttributesReader.APPLICATION_NAME))
        assertEquals(reader.getApplicationName(), APP_NAME);
    if(!Objects.equals(except, AttributesReader.APPLICATION_INSTANCE_ID))
        assertEquals(reader.getApplicationInstanceId(), APP_INSTANCE_ID);
    if(!Objects.equals(except, AttributesReader.APPLICATION_LANGUAGE))
        assertEquals(reader.getApplicationLanguage(), APP_LANG);
    if(!Objects.equals(except, AttributesReader.METHOD_FQN))
      assertEquals(reader.getMethodFqn(), FQN);
    
    // k8s stuff
    if(!Objects.equals(except, AttributesReader.K8S_POD_NAME))
        assertEquals(reader.getK8sPodName(), K8S_POD_NAME);
    if(!Objects.equals(except, AttributesReader.K8S_NAMESPACE_NAME))
        assertEquals(reader.getK8sNamespace(), K8S_NAMESPACE);
    if(!Objects.equals(except, AttributesReader.K8S_DEPLOYMENT_NAME))
        assertEquals(reader.getK8sDeploymentName(), K8S_DEPLOYMENT);
    if(!Objects.equals(except, AttributesReader.K8S_NODE_NAME))
        assertEquals(reader.getK8sNodeName(), K8S_NODE);

  }

  private List<KeyValue> generateValidAttributesMap() {
    List<KeyValue> attributes = new ArrayList<>();

    attributes.add(KeyValue.newBuilder().setKey(KEY_LANDSCAPE_TOKEN)
        .setValue(AnyValue.newBuilder().setStringValue(TOKEN).build()).build());

    attributes.add(KeyValue.newBuilder().setKey(KEY_LANDSCAPE_SECRET)
        .setValue(AnyValue.newBuilder().setStringValue(SECRET).build()).build());

    attributes.add(KeyValue.newBuilder().setKey(KEY_HOST_NAME)
        .setValue(AnyValue.newBuilder().setStringValue(HOSTNAME).build()).build());

    attributes.add(KeyValue.newBuilder().setKey(KEY_HOST_IP)
        .setValue(AnyValue.newBuilder().setStringValue(HOST_IP).build()).build());

    attributes.add(KeyValue.newBuilder().setKey(KEY_APPLICATION_NAME)
        .setValue(AnyValue.newBuilder().setStringValue(APP_NAME).build()).build());

    attributes.add(KeyValue.newBuilder().setKey(KEY_APPLICATION_INSTANCE_ID)
        .setValue(AnyValue.newBuilder().setStringValue(APP_INSTANCE_ID).build()).build());

    attributes.add(KeyValue.newBuilder().setKey(KEY_APPLICATION_LANGUAGE)
        .setValue(AnyValue.newBuilder().setStringValue(APP_LANG).build()).build());

    attributes.add(KeyValue.newBuilder().setKey(KEY_METHOD_FQN)
        .setValue(AnyValue.newBuilder().setStringValue(FQN).build()).build());

    attributes.add(newKeyValueString(AttributesReader.K8S_POD_NAME, K8S_POD_NAME));
    attributes.add(newKeyValueString(AttributesReader.K8S_NAMESPACE_NAME, K8S_NAMESPACE));
    attributes.add(newKeyValueString(AttributesReader.K8S_DEPLOYMENT_NAME, K8S_DEPLOYMENT));
    attributes.add(newKeyValueString(AttributesReader.K8S_NODE_NAME, K8S_NODE));

    return attributes;
  }

  private Span generateSpanFromAttributesMap(List<KeyValue> attributes) {
    return Span.newBuilder()
        .setTraceId(
            ByteString.copyFrom("50c246ad9c9883d1558df9f19b9ae7a6", Charset.defaultCharset()))
        .setSpanId(ByteString.copyFrom("7ef83c66eabd5fbb", Charset.defaultCharset()))
        .setParentSpanId(ByteString.copyFrom("7ef83c66efe42aaa", Charset.defaultCharset()))
        .setStartTimeUnixNano(1667986986000L)
        .setEndTimeUnixNano(1667987046000L)
        .addAllAttributes(attributes).build();
  }

  private List<KeyValue> removeElementAndReturnAttributesMap(String keyToBeRemoved,
      List<KeyValue> attributes) {

    List<KeyValue> resultList = new ArrayList<>();

    for (KeyValue keyVal : attributes) {
      if (keyVal.getKey().equals(keyToBeRemoved)) {
        // do nothing
      } else {
        resultList.add(keyVal);
      }
    }

    return resultList;
  }


  // TODO: dont do code duplication with StrictValidatorTest.newKeyValueString
  // this should be done when replacing all 
  // KeyValue.newBuilder().setKey(KEY_HOST_IP).setValue(AnyValue.newBuilder().setStringValue(HOST_IP).build()).build()
  // with this method.
  // i am not doing this here, because i am implementing a specific feature; not refactoring the whole project.
  // so pwease forgive me, uwu
  //
  // i am not quite sure if this sounds condescending.
  // if yes: it's not meant to be.
  // i cant tell since im autistic *jazz hands*
  public static KeyValue newKeyValueString(String key, String value){
    return KeyValue.newBuilder().setKey(key)
        .setValue(AnyValue.newBuilder().setStringValue(value)).build();
  }

}

File: src/test/java/net/explorviz/adapter/service/converter/AttributesReaderTest.kt
package net.explorviz.adapter.service.converter

import com.google.protobuf.ByteString
import io.opentelemetry.proto.common.v1.AnyValue
import io.opentelemetry.proto.common.v1.KeyValue
import io.opentelemetry.proto.trace.v1.Span
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test

class AttributesReaderTest {

    companion object {
        private const val KEY_LANDSCAPE_TOKEN = AttributesReader.LANDSCAPE_TOKEN
        private const val KEY_LANDSCAPE_SECRET = AttributesReader.TOKEN_SECRET
        private const val KEY_GIT_COMMIT_CHECKSUM = AttributesReader.GIT_COMMIT_CHECKSUM
        private const val KEY_HOST_NAME = AttributesReader.HOST_NAME
        private const val KEY_HOST_IP = AttributesReader.HOST_IP
        private const val KEY_APPLICATION_NAME = AttributesReader.APPLICATION_NAME
        private const val KEY_APPLICATION_INSTANCE_ID = AttributesReader.APPLICATION_INSTANCE_ID
        private const val KEY_APPLICATION_LANGUAGE = AttributesReader.APPLICATION_LANGUAGE
        private const val KEY_METHOD_FQN = AttributesReader.METHOD_FQN
        private const val KEY_K8S_POD_NAME = AttributesReader.K8S_POD_NAME
        private const val KEY_K8S_NAMESPACE = AttributesReader.K8S_NAMESPACE_NAME
        private const val KEY_K8S_DEPLOYMENT = AttributesReader.K8S_DEPLOYMENT_NAME
        private const val KEY_K8S_NODE = AttributesReader.K8S_NODE_NAME

        private const val TOKEN = "tok"
        private const val SECRET = "secret"
        private const val GIT_COMMIT_CHECKSUM = "gitCommitChecksum"
        private const val HOSTNAME = "Host"
        private const val HOST_IP = "1.2.3.4"
        private const val APP_NAME = "Test App"
        private const val APP_INSTANCE_ID = "1234L"
        private const val APP_LANG = "java"
        private const val FQN = "foo.bar.test()"
        private const val NAME = "span name"
        private const val K8S_POD_NAME = "my name is pod. james pod."
        private const val K8S_NAMESPACE = "name "
        private const val K8S_DEPLOYMENT = "deployment name"
        private const val K8S_NODE = "node.js"
    }

    @Test
    fun testValidSpanReadOut() {
        val attr = generateValidAttributesMap()
        val span = generateSpanFromAttributesMap(attr)
        val reader = AttributesReader(span)

        assertEquals(reader.landscapeToken, TOKEN)
        assertEquals(reader.secret, SECRET)
        assertEquals(reader.gitCommitChecksum, GIT_COMMIT_CHECKSUM)
        assertEquals(reader.hostName, HOSTNAME)
        assertEquals(reader.hostIpAddress, HOST_IP)
        assertEquals(reader.applicationName, APP_NAME)
        assertEquals(reader.applicationInstanceId, APP_INSTANCE_ID)
        assertEquals(reader.applicationLanguage, APP_LANG)
        assertEquals(reader.methodFqn, FQN)
        assertExcept(reader, null)
    }

    @Test
    fun testDefaultTokenReadOut() {
        var attr = generateValidAttributesMap()
        attr = removeElementAndReturnAttributesMap(KEY_LANDSCAPE_TOKEN, attr)
        val span = generateSpanFromAttributesMap(attr)
        val reader = AttributesReader(span)

        assertExcept(reader, AttributesReader.LANDSCAPE_TOKEN)
        assertEquals(reader.landscapeToken, DefaultAttributeValues.DEFAULT_LANDSCAPE_TOKEN)
    }

    @Test
    fun testDefaultSecretReadOut() {
        var attr = generateValidAttributesMap()
        attr = removeElementAndReturnAttributesMap(KEY_LANDSCAPE_SECRET, attr)
        val span = generateSpanFromAttributesMap(attr)
        val reader = AttributesReader(span)

        assertExcept(reader, AttributesReader.TOKEN_SECRET)
        assertEquals(reader.secret, DefaultAttributeValues.DEFAULT_LANDSCAPE_SECRET)
    }

    @Test
    fun testDefaultHostnameReadOut() {
        var attr = generateValidAttributesMap()
        attr = removeElementAndReturnAttributesMap(KEY_HOST_NAME, attr)
        val span = generateSpanFromAttributesMap(attr)
        val reader = AttributesReader(span)

        assertExcept(reader, AttributesReader.HOST_NAME)
        assertEquals(reader.hostName, DefaultAttributeValues.DEFAULT_HOST_NAME)
    }

    @Test
    fun testDefaultHostIpReadOut() {
        var attr = generateValidAttributesMap()
        attr = removeElementAndReturnAttributesMap(KEY_HOST_IP, attr)
        val span = generateSpanFromAttributesMap(attr)
        val reader = AttributesReader(span)

        assertExcept(reader, AttributesReader.HOST_IP)
        assertEquals(reader.hostIpAddress, DefaultAttributeValues.DEFAULT_HOST_IP)
    }

    @Test
    fun testDefaultAppNameReadOut() {
        var attr = generateValidAttributesMap()
        attr = removeElementAndReturnAttributesMap(KEY_APPLICATION_NAME, attr)
        val span = generateSpanFromAttributesMap(attr)
        val reader = AttributesReader(span)

        assertExcept(reader, AttributesReader.APPLICATION_NAME)
        assertEquals(reader.applicationName, DefaultAttributeValues.DEFAULT_APP_NAME)
    }

    @Test
    fun testDefaultAppInstanceIdReadOut() {
        var attr = generateValidAttributesMap()
        attr = removeElementAndReturnAttributesMap(KEY_APPLICATION_INSTANCE_ID, attr)
        val span = generateSpanFromAttributesMap(attr)
        val reader = AttributesReader(span)

        assertExcept(reader, AttributesReader.APPLICATION_INSTANCE_ID)
        assertEquals(reader.applicationInstanceId, DefaultAttributeValues.DEFAULT_APP_INSTANCE_ID)
    }

    @Test
    fun testDefaultAppLangReadOut() {
        var attr = generateValidAttributesMap()
        attr = removeElementAndReturnAttributesMap(KEY_APPLICATION_LANGUAGE, attr)
        val span = generateSpanFromAttributesMap(attr)
        val reader = AttributesReader(span)

        assertExcept(reader, AttributesReader.APPLICATION_LANGUAGE)
        assertEquals(reader.applicationLanguage, DefaultAttributeValues.DEFAULT_APP_LANG)
    }

    @Test
    fun testDefaultFqnReadOut() {
        var attr = generateValidAttributesMap()
        attr = removeElementAndReturnAttributesMap(KEY_METHOD_FQN, attr)
        val span = generateSpanFromAttributesMap(attr)
        val reader = AttributesReader(span)

        assertExcept(reader, AttributesReader.METHOD_FQN)
        assertEquals(reader.methodFqn, DefaultAttributeValues.DEFAULT_CLASS_FQN + "." + span.getName())
    }

    @Test
    fun testFqnFromNameReadOut() {
        val outerPackage = "net"
        val innerPackage = "explorviz"
        val className = "Reader"
        val methodName = "someNiceMethod()"

        // Only method name given
        var span = Span.newBuilder().setName(methodName).build()
        var reader = AttributesReader(span)
        assertEquals(DefaultAttributeValues.DEFAULT_CLASS_FQN + "." + methodName, reader.methodFqn)

        // Class and method given
        span = Span.newBuilder().setName("$className.$methodName").build()
        reader = AttributesReader(span)
        assertEquals("${DefaultAttributeValues.DEFAULT_PACKAGE_NAME}.${className}.${methodName}", reader.methodFqn)

        // Class and package given
        span = Span.newBuilder().setName("$innerPackage.$className.$methodName").build()
        reader = AttributesReader(span)
        assertEquals("$innerPackage.$className.$methodName", reader.methodFqn)

        // Class and two packages given
        span = Span.newBuilder().setName("$outerPackage.$innerPackage.$className.$methodName").build()
        reader = AttributesReader(span)
        assertEquals("$outerPackage.$innerPackage.$className.$methodName", reader.methodFqn)
    }

    private fun assertExcept(reader: AttributesReader, except: String?) {
        if (except != AttributesReader.LANDSCAPE_TOKEN) {
            assertEquals(reader.landscapeToken, TOKEN)
        }
        if (except != AttributesReader.TOKEN_SECRET) {
            assertEquals(reader.secret, SECRET)
        }
        if (except != AttributesReader.HOST_NAME) {
            assertEquals(reader.hostName, HOSTNAME)
        }
        if (except != AttributesReader.HOST_IP) {
            assertEquals(reader.hostIpAddress, HOST_IP)
        }
        if (except != AttributesReader.APPLICATION_NAME) {
            assertEquals(reader.applicationName, APP_NAME)
        }
        if (except != AttributesReader.APPLICATION_INSTANCE_ID) {
            assertEquals(reader.applicationInstanceId, APP_INSTANCE_ID)
        }
        if (except != AttributesReader.APPLICATION_LANGUAGE) {
            assertEquals(reader.applicationLanguage, APP_LANG)
        }
        if (except != AttributesReader.METHOD_FQN) {
            assertEquals(reader.methodFqn, FQN)
        }

        // Kubernetes-related checks
        if (except != AttributesReader.K8S_POD_NAME) {
            assertEquals(reader.k8sPodName, K8S_POD_NAME)
        }
        if (except != AttributesReader.K8S_NAMESPACE_NAME) {
            assertEquals(reader.k8sNamespace, K8S_NAMESPACE)
        }
        if (except != AttributesReader.K8S_DEPLOYMENT_NAME) {
            assertEquals(reader.k8sDeploymentName, K8S_DEPLOYMENT)
        }
        if (except != AttributesReader.K8S_NODE_NAME) {
            assertEquals(reader.k8sNodeName, K8S_NODE)
        }
    }

    private fun generateValidAttributesMap(): MutableList<KeyValue> {
        return mutableListOf(
            KeyValue.newBuilder()
                .setKey(KEY_LANDSCAPE_TOKEN)
                .setValue(AnyValue.newBuilder().setStringValue(TOKEN).build())
                .build(),
            KeyValue.newBuilder()
                .setKey(KEY_LANDSCAPE_SECRET)
                .setValue(AnyValue.newBuilder().setStringValue(SECRET).build())
                .build(),
            KeyValue.newBuilder()
                .setKey(KEY_GIT_COMMIT_CHECKSUM)
                .setValue(AnyValue.newBuilder().setStringValue(GIT_COMMIT_CHECKSUM).build())
                .build(),
            KeyValue.newBuilder()
                .setKey(KEY_HOST_NAME)
                .setValue(AnyValue.newBuilder().setStringValue(HOSTNAME).build())
                .build(),
            KeyValue.newBuilder()
                .setKey(KEY_HOST_IP)
                .setValue(AnyValue.newBuilder().setStringValue(HOST_IP).build())
                .build(),
            KeyValue.newBuilder()
                .setKey(KEY_APPLICATION_NAME)
                .setValue(AnyValue.newBuilder().setStringValue(APP_NAME).build())
                .build(),
            KeyValue.newBuilder()
                .setKey(KEY_APPLICATION_INSTANCE_ID)
                .setValue(AnyValue.newBuilder().setStringValue(APP_INSTANCE_ID).build())
                .build(),
            KeyValue.newBuilder()
                .setKey(KEY_APPLICATION_LANGUAGE)
                .setValue(AnyValue.newBuilder().setStringValue(APP_LANG).build())
                .build(),
            KeyValue.newBuilder()
                .setKey(KEY_METHOD_FQN)
                .setValue(AnyValue.newBuilder().setStringValue(FQN).build())
                .build(),
            KeyValue.newBuilder()
                .setKey(KEY_K8S_POD_NAME)
                .setValue(AnyValue.newBuilder().setStringValue(K8S_POD_NAME).build())
                .build(),
            KeyValue.newBuilder()
                .setKey(KEY_K8S_NAMESPACE)
                .setValue(AnyValue.newBuilder().setStringValue(K8S_NAMESPACE).build())
                .build(),
            KeyValue.newBuilder()
                .setKey(KEY_K8S_DEPLOYMENT)
                .setValue(AnyValue.newBuilder().setStringValue(K8S_DEPLOYMENT).build())
                .build(),
            KeyValue.newBuilder()
                .setKey(KEY_K8S_NODE)
                .setValue(AnyValue.newBuilder().setStringValue(K8S_NODE).build())
                .build()
        )
    }

    private fun removeElementAndReturnAttributesMap(key: String, map: MutableList<KeyValue>): MutableList<KeyValue> {
        return map.filter { it.key != key }.toMutableList()
    }

    private fun generateSpanFromAttributesMap(attributes: MutableList<KeyValue>): Span {
        return Span.newBuilder()
            .setTraceId(ByteString.copyFrom("50c246ad9c9883d1558df9f9b9ae7a6", Charsets.UTF_8))
            .setSpanId(ByteString.copyFrom("7ef83c66eabd5fbb", Charsets.UTF_8))
            .setParentSpanId(ByteString.copyFrom("7ef83c66efe42aaa", Charsets.UTF_8))
            .setName(NAME)
            .setStartTimeUnixNano(1667986986000L)
            .setEndTimeUnixNano(1667987046000L)
            .addAllAttributes(attributes)
            .build()
    }
}
